Modules/_threadmodule.c

```diff
@@ -1108,42 +1169,50 @@ rlock_release(PyObject *op, PyObject *Py_UNUSED(dummy))
     Py_RETURN_NONE;
 }
 
-PyDoc_STRVAR(rlock_release_doc,
-"release($self, /)\n\
---\n\
-\n\
-Release the lock, allowing another thread that is blocked waiting for\n\
-the lock to acquire the lock.  The lock must be in the locked state,\n\
-and must be locked by the same thread that unlocks it; otherwise a\n\
-`RuntimeError` is raised.\n\
-\n\
-Do note that if the lock was acquire()d several times in a row by the\n\
-current thread, release() needs to be called as many times for the lock\n\
-to be available for other threads.");
+/*[clinic input]
+_thread.RLock.__exit__
+    exc_type: object = None
+    exc_value: object = None
+    exc_tb: object = None
+    /
 
-PyDoc_STRVAR(rlock_exit_doc,
-"__exit__($self, /, *exc_info)\n\
---\n\
-\n\
-Release the lock.");
+Release the lock.
+
+[clinic start generated code]*/
 
 static PyObject *
-rlock_locked(PyObject *op, PyObject *Py_UNUSED(ignored))
+_thread_RLock___exit___impl(rlockobject *self, PyObject *exc_type,
+                            PyObject *exc_value, PyObject *exc_tb)
+/*[clinic end generated code: output=79bb44d551aedeb5 input=99d5cb8c7cef3077]*/
+{
+    return _thread_RLock_release_impl(self);
+}
+
+/*[clinic input]
+_thread.RLock.locked
+
+Return a boolean indicating whether this object is locked right now.
+[clinic start generated code]*/
+
+static PyObject *
+_thread_RLock_locked_impl(rlockobject *self)
+/*[clinic end generated code: output=e9b6060492b3f94e input=8866d9237ba5391b]*/
 {
-    rlockobject *self = rlockobject_CAST(op);
     int is_locked = rlock_locked_impl(self);
     return PyBool_FromLong(is_locked);
 }
 
-PyDoc_STRVAR(rlock_locked_doc,
-"locked()\n\
-\n\
-Return a boolean indicating whether this object is locked right now.");
+/*[clinic input]
+_thread.RLock._acquire_restore
+    *args: tuple
```

Comment:
Why not use Argument Clinic to parse argument?
```suggestion
    count: Py_ssize_t
    owner: PyThread_ident_t
    /
```
You simply needs to add a custom converter for PyThread_ident_t.

Reply:
I tried that but that didn't work.
```c
    if (!PyArg_ParseTuple(args, "(n" Py_PARSE_THREAD_IDENT_T "):_acquire_restore",
            &count, &owner))
        return NULL;
```

The count and owner is actually passed as tuple here and I don't think AC can parse such case.

Reply:
Then it should bye
```suggestion
    count_owner: object
    /
```
or
```suggestion
    count_owner: object(subclass_of="&PyTuple_Type")
    /
```
And if it is internal function, it would be nice to change it to accept two separate arguments instead of a pair in the following PR.

Reply:
I don't think using `count_owner` is any better than just `*args:tuple`. How about keeping this as it is and changing is to use two args in a different PR?

Reply:
This is purely an internal function, does it matter?

Reply:
Oh, nevermind. This method is exposed to `threading` users:

https://github.com/python/cpython/blob/75d20b2c67e91dd468e7f5fe4ac465e6d32f8c09/Lib/threading.py#L296-L297

The diff from this PR is the following:

```diff
3c3
< _acquire_restore(self, state, /) unbound _thread.RLock method
---
> _acquire_restore(self, /, *args) unbound _thread.RLock method
```

I think we should just switch back to a single `state` argument here.

Reply:
Hmm, changing it to use state as single arg leads to test failures in `test_notify`
```console
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/kumaraditya/work/cpython/Lib/test/lock_tests.py", line 786, in f
    |     result = cond.wait_for(lambda : state==4, timeout=0.1)
    |   File "/Users/kumaraditya/work/cpython/Lib/threading.py", line 397, in wait_for
    |     self.wait(waittime)
    |     ~~~~~~~~~^^^^^^^^^^
    |   File "/Users/kumaraditya/work/cpython/Lib/threading.py", line 371, in wait
    |     self._acquire_restore(saved_state)
    |     ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
    | TypeError: _acquire_restore() takes exactly 1 argument (2 given)
    | 
```

I'll need to take a deeper look into this to understand it

Reply:
This method currently uses METH_VARARGS unnecessary. This works, but it is confusing. This PR reproduces weirdness of the current code.

It can be fixed by a patch:
```patch
diff --git a/Modules/_threadmodule.c b/Modules/_threadmodule.c
index f82ad6870f8..41c76bd3b6a 100644
--- a/Modules/_threadmodule.c
+++ b/Modules/_threadmodule.c
@@ -1141,13 +1141,13 @@ PyDoc_STRVAR(rlock_locked_doc,
 Return a boolean indicating whether this object is locked right now.");
 
 static PyObject *
-rlock_acquire_restore(PyObject *op, PyObject *args)
+rlock_acquire_restore(PyObject *op, PyObject *state)
 {
     rlockobject *self = rlockobject_CAST(op);
     PyThread_ident_t owner;
     Py_ssize_t count;
 
-    if (!PyArg_ParseTuple(args, "(n" Py_PARSE_THREAD_IDENT_T "):_acquire_restore",
+    if (!PyArg_Parse(state, "(n" Py_PARSE_THREAD_IDENT_T "):_acquire_restore",
             &count, &owner))
         return NULL;
 
@@ -1276,7 +1276,7 @@ static PyMethodDef rlock_methods[] = {
     {"_is_owned",     rlock_is_owned,
      METH_NOARGS, rlock_is_owned_doc},
     {"_acquire_restore", rlock_acquire_restore,
-     METH_VARARGS, rlock_acquire_restore_doc},
+     METH_O, rlock_acquire_restore_doc},
     {"_release_save", rlock_release_save,
      METH_NOARGS, rlock_release_save_doc},
     {"_recursion_count", rlock_recursion_count,
```
Or we can simply incorporate these changes in this PR. At least it will generate better signature.

Reply:
> Hmm, changing it to use state as single arg leads to test failures in `test_notify`

Because you also need to change `PyArg_ParseTuple` to `
PyArg_Parse`.

Reply:
Ah, sorry I missed that part.

Reply:
`test_notify` in `test_threading`? All passed successfully.

Reply:
> test_notify in test_threading? All passed successfully.

Yes it passed now, I missed changing `PyArg_ParseTuple` to `PyArg_Parse`.